tests/std/tests/P0067R5_charconv/test.cpp

```diff
@@ -212,39 +212,39 @@ constexpr const char* output_max_digit[] = {"skip0", "skip1", "11", "12", "13",
 // https://www.wolframalpha.com : Table[BaseForm[k, n], {k, {MEOW, MEOW, MEOW}}, {n, 2, 36}]
 constexpr uint64_t stress_chunks_positive                          = 12000000345000678900ULL;
 constexpr pair<uint64_t, array<const char*, 37>> output_positive[] = {
-    {123, {{"skip0", "skip1", "1111011", "11120", "1323", "443", "323", "234", "173", "146", "123", "102", "a3", "96",
-              "8b", "83", "7b", "74", "6f", "69", "63", "5i", "5d", "58", "53", "4n", "4j", "4f", "4b", "47", "43",
-              "3u", "3r", "3o", "3l", "3i", "3f"}}},
-    {INT8_MAX, {{"skip0", "skip1", "1111111", "11201", "1333", "1002", "331", "241", "177", "151", "127", "106", "a7",
-                   "9a", "91", "87", "7f", "78", "71", "6d", "67", "61", "5h", "5c", "57", "52", "4n", "4j", "4f", "4b",
-                   "47", "43", "3v", "3s", "3p", "3m", "3j"}}},
-    {161, {{"skip0", "skip1", "10100001", "12222", "2201", "1121", "425", "320", "241", "188", "161", "137", "115",
-              "c5", "b7", "ab", "a1", "98", "8h", "89", "81", "7e", "77", "70", "6h", "6b", "65", "5q", "5l", "5g",
-              "5b", "56", "51", "4t", "4p", "4l", "4h"}}},
+    {123U, {{"skip0", "skip1", "1111011", "11120", "1323", "443", "323", "234", "173", "146", "123", "102", "a3", "96",

```

Comment:
I'm suspicious that these changes suggest this PR will break lots of user code. What's going on here?

Reply:
After std::pair's constructor changes I got the warnings:
```
test.cpp(313): note: see reference to function template instantiation 'std::pair<uint64_t,std::array<const char *,37>>::pair<int,std::array<const char *,37>,0>(_Other1 &&,_Other2 &&) noexcept' being compiled
        with
        [
            _Other1=int,
            _Other2=std::array<const char *,37>
        ]
utility(193): warning C4365: 'initializing': conversion from '_Ty' to '_Ty1', signed/unsigned mismatch
        with
        [
            _Ty=int
        ]
        and
        [
            _Ty1=uint64_t
        ]
```
So I fixed the warning by changing the type of integers to unsigned type.

Reply:
> will break lots of user code

Only the code of those who wanted it (who set the `/WX`)

Reply:
Semi-reduced to:
```c++
// Compile with cl /nologo /c /w14365
#include <array>

template <class T, class U>
struct pair {
    T first;
    U second;

    constexpr pair(const T& t, const U& u) : first(t), second(u) {}

#ifdef BEFORE
    template <class V, class W> // Ok
#else
    template <class V = T, class W = U> // Diagnostics on line 17
#endif
    constexpr pair(V&& v, W&& w) : 
        first(static_cast<V&&>(v)), second(static_cast<W&&>(w)) {} // Line 17
};

using int64 = long long;
using uint64 = unsigned long long;

constexpr pair<uint64, std::array<const char*, 1>> output_positive[] = {
    {char{0x7f}, {{"skip0"}}},
    {short{0x7fff}, {{"skip0"}}},
    {int{0x7fff'ffff}, {{"skip0"}}},
    {int64{0x7fff'ffff'ffff'ffff}, {{"skip0"}}},
};
```
I'm not certain, but I _think_ there should be no conversion/narrowing diagnostics here since we're converting constant expressions whose values are all representable by the target type (`uint64_t`). In any case, the presence or absence of the default template arguments should affect whether or not a conversion/narrowing warning is emitted.

In any case, I think we're fine to merge this as-is now that I've filed the above repro as a compiler diagnostic bug (DevCom-1458942). Hopefully this pattern is rare enough that we won't see diagnostic regressions in user code and/or the compiler will get this cleaned up before it happens to users.

Reply:
This sounds like the difference is in _where_ the conversion happens. With `-DBEFORE`, deduction for `W` fails so `pair(const T& t, const U& u)` is called and the conversion happens outside the constructor. Without it, the constructor template is called since it is a better match, and the conversion happens inside it instead.